internal/sys/file_table.go

```diff
@@ -0,0 +1,141 @@
+//
+// Key generation is managed by the table, and currently uses a strategy similar
+// to the file descriptor allocation on unix systems: the lowest key not mapped
+// to any value is used when inserting a new item.
+//
+// The data structure is used to implement the file descriptor table of the WASI
+// implementation of Wazero. The WASI standard documents that programs cannot
+// expect that file descriptor numbers will be allocated with a lowest-first
+// strategy (like it is done on unix systems), and they shouldi instead assume
+// the values will be random. Since "random" is a very imprecise concept in
+// computers, we technically satisfying the implementation with the key
+// allocation strategy we use here. We could imagine adding more "randomness"
+// to the key selection process, however this should never be used as a security
+// measure to prevent applications from guessing the next file number so there
+// are no strong incentives to complicate the logic.
+//
+// The data structure optimizes for memory density and lookup performance,
+// trading off compute at insertion time. This is a useful compromise for the
+// use cases we employ it with: files are usually read or written a lot more
+// often than they are opened, each operation requires a table lookup so we are
+// better off spending extra compute to insert files in the table in order to
+// get cheaper lookups. Memory efficiency is also crucial to support scaling
+// with programs that open thousands of files: having a high or non-linear
+// memory-to-item ratio could otherwise be used as an attack vector by malicous
+// applications attempting to damage performance of the host.
+type FileTable struct {
+	masks []uint64
+	files []*FileEntry
+}
+
+// Len returns the number of files stored in the table.
+func (t *FileTable) Len() (n int) {
+	// We could make this a O(1) operation if we cached the number of files in
+	// the table. More state usually means more problems, so until we have a
+	// clear need for this, the simple implementation may be a better trade off.
+	for _, mask := range t.masks {
+		n += bits.OnesCount64(mask)
+	}
+	return n
+}
+
+// Grow ensures that t has enough room for n files, potentially reallocating the
+// internal buffers if their capacity was too small to hold this many files.
+func (t *FileTable) Grow(n int) {
+	if n = (n*8 + 7) / 8; n > len(t.masks) {
+		masks := make([]uint64, n)
+		copy(masks, t.masks)
+
+		files := make([]*FileEntry, n*64)
+		copy(files, t.files)
+
+		t.masks = masks
+		t.files = files
+	}
+}
+
+// Insert inserts the given file to the table, returning the fd that it is
+// mapped to.
+//
+// The method does not perform deduplication, it is possible for the same file
+// to be inserted multiple times, each insertion will return a different fd.
+func (t *FileTable) Insert(file *FileEntry) (fd uint32) {
+	offset := 0
+insert:
+	// TODO: this loop could be made a lot more efficient using vectorized
```

Comment:
does go support vectorized operations now or in the near future? If so, maybe not a TODO

Reply:
I'll change this to just a note rather than a TODO.